Validate data

andrayantelo opened the pull request:
Working on validation rules as well as fixing read and write rules. Addresses #117

andrayantelo committed:
trying to figure out validation rules

andrayantelo committed:
Added validation rules and modified write rules
Added validation rules for startDateString, endDateString, and title to see if I can do away with the validateInput method. Modified writers/ .write permissions so that if the calendar in question is in the user/'s allCalendarIds then they are allowed to write in it. It used to be that they could only write to writers if they were already in writers... but they aren't in writers to begin with when creating a calendar.

andrayantelo committed:
fixed the save method in the firebase store object

andrayantelo committed:
updated save method in the firebase object so that permission is granted to user

johntyree commented on docs/store.js:
```diff
@@ -91,13 +91,28 @@ var firebaseCalendarStorage = function(params) {
-        // Store the user's allCalendarIds and their calendarState .. setting the
-        // active calendar has its own method.
+        // Store the calendar into allCalendarIds first so that the user
+        // has permission to write into calendars/calUniqueId
```
Maybe you should be allowed to create a new calendar as long as the final state is that the calendar is also in your users list of cal ids?

andrayantelo committed:
whatever

andrayantelo committed:
added a TODO

andrayantelo committed:
Added a separate method to write to writers dir
ALso, made sure that addtoWriter's get's called when a calendar is created.

andrayantelo committed:
added a removeUser method

andrayantelo committed:
remove contents of dropdown before filling it when a user signs in

andrayantelo committed:
clear dropdown inside of fillDropdown instead of onAuthChanged

andrayantelo committed:
wrote a checkit method that updates user information on the page

andrayantelo committed:
Added addReader and removeReader methods
Also made sure to add the user to the readers list when they create a calendar.

andrayantelo committed:
updating rules to include readers dir

andrayantelo committed:
updated the database rules for writers and readers directories

andrayantelo committed:
added validation rules

johntyree commented on docs/store.js:
```diff
@@ -91,10 +91,13 @@ var firebaseCalendarStorage = function(params) {
+        // states, etc.
         var updates = {};
         updates['users/' + userId + '/allCalendarIds/' + calUniqueId] = calTitle;
+        updates['calendars/' + calUniqueId + '/calendarState'] = calState;
```
Not important, but you can also just define all these when you define the object literal. Not necessarily better or worse.

```
var updates = {
  'users/' + userId + '/allCalendarIds/' + calUniqueId: calTitle,
 ....
}
```

andrayantelo replied:
Oh okay. Will keep that in mind.

andrayantelo committed:
Edited uniqueId generator so that the uniqueIds are strings

andrayantelo committed:
updated sample json

andrayantelo committed:
wrote validation rules for calendarState values

andrayantelo committed:
added validation, read, and write rules for the data under the user dir: allCalendarIds and currentActiveCalendar

andrayantelo committed:
moving read and write rules around. Fixed permissions for currentActiveCalendar

andrayantelo committed:
Fixed data validation rules for readers and writers

andrayantelo committed:
fixed some validation rules, added endWork to setActiveById method, added some comments

andrayantelo committed:
Wrote a initializeCalendar method
This method is for the firebase store object. Wrote it because we need to have the user added to the
database's writers directory before being able to write to any of the other nodes in the database.
Still needs testing.

andrayantelo committed:
fixed syntax errors

johntyree commented on database.rules.json:
```diff
@@ -3,24 +3,57 @@
         ".read" : "auth !== null && auth.uid === $uid",
-        ".write" : "auth !== null && auth.uid === $uid"
+        "allCalendarIds" : {
+          //".write" : "true",
```
Remove this

andrayantelo replied:
Done.

johntyree commented on database.rules.json:
```diff
@@ -3,24 +3,57 @@
+          }
+        },
+        "currentActiveCalendar" : {
+            ".write" : "auth !== null && auth.uid === $uid && (!data.exists() || root.child('calendars/' + newData.val() + '/writers/' + auth.uid).exists())",  
```
Really you only need to be able to read the calendar to show it.

johntyree replied:
Not important now, but if it's supposed to be possible to share a calendar somehow then we'll have to look at this again

andrayantelo replied:
Okay.

andrayantelo replied:
Changed it to readers.

johntyree commented on database.rules.json:
```diff
@@ -3,24 +3,57 @@
+          ".read" : "auth !== null && (!data.exists() || root.child('calendars/' + $calendarId + '/readers/' + auth.uid).exists())",
+          ".write" : "auth !== null && (!data.exists() || root.child('calendars/' +$calendarId + '/writers/' + auth.uid).exists())",
+          "startDateString": {
+            ".validate" : "newData.isString() && newData.val().length > 0 && newData.val().matches(/\\d{8}/)"
```
I think the regex alone is enough here. If it's not a string it will fail. If it is, then the length will be restricted to exactly 8.

andrayantelo replied:
Left just the regex.

johntyree commented on database.rules.json:
```diff
@@ -3,24 +3,57 @@
+            ".validate": "newData.isString() && newData.val().length > 0 && newData.val().length < 256"
+          },
+          "uniqueId" : {
+            ".validate" : "newData.isString() && newData.val().length > 0"
```
Probably something like `newData.val().matches(/([A-Za-z0-9]+/)` is better here. The id should be just be some blob of alphanumeric something or another.

andrayantelo replied:
Used that as uniqueId's validation rule.

johntyree commented on database.rules.json:
```diff
@@ -3,24 +3,57 @@
         "writers" : {
-          ".read" : "auth !== null && data.child(auth.uid).exists()",
-          ".write" : "auth !== null && data.child(auth.uid).exists()"
+          ".write" : "auth !== null && (!data.exists() || root.child('calendars/' + $calendarId + '/writers/' + auth.uid).exists())",
```
You're already here.  `data.child(auth.uid).exists()` ?  Maybe `data.parent.child` at worst.

andrayantelo replied:
Changed it to the first option.

johntyree commented on docs/store.js:
```diff
@@ -81,6 +81,69 @@ var firebaseCalendarStorage = function(params) {
+        
+        self.startWork();
+        return self.database.ref().update(updates)
+        .then(function() {
```
fix the indentation here please

andrayantelo replied:
Fixed.

johntyree commented on docs/store.js:
```diff
@@ -81,6 +81,69 @@ var firebaseCalendarStorage = function(params) {
+        // Remove a user form the writers directory in the database
+        
+        self.database.ref('calendars/' + calObj.state.uniqueId + '/writers/' + user.uid).remove()
+        .then(function() {
```
indentation on all of these.

andrayantelo replied:
Fixed.

johntyree commented on docs/store.js:
```diff
@@ -91,24 +154,24 @@ var firebaseCalendarStorage = function(params) {
-        // Store the user's allCalendarIds and their calendarState .. setting the
-        // active calendar has its own method.
+        // Store the calendar into allCalendarIds, store calState
+        // TODO work on save method, how to save readers, writers, permissions, 
```
Is this out of date now?

andrayantelo replied:
Yes. Removed.

johntyree commented on docs/store.js:
```diff
@@ -230,6 +294,26 @@ var firebaseCalendarStorage = function(params) {
+            self.addReader(self.user, calendarObj);
+            self.setActiveById(calendarObj.state.uniqueId);
+            self.save(calendarObj);
+            self.endWork();
```
You're not done with work until those other promises return.  You need a `Promise.all` here.

andrayantelo replied:
Done.

andrayantelo committed:
fixed some indentation in store js and fixed validation rules in rules file

andrayantelo committed:
Fixed initializeCalendar method in firebase store object to wait until all promises are resolved before running endWork

johntyree commented on docs/store.js:
```diff
@@ -210,21 +207,21 @@ var firebaseCalendarStorage = function(params) {
             })
             .catch(function(err) {
-                console.error('Unable to remove calendar: ' + err);
+                console.error('Unable to remove calendar from user\'s allCalendarIds: ' + err);
```
It's usually easier to read if you make use of the other type of quotes so that you don't have to escape.

andrayantelo replied:
Do you mean back ticks?

johntyree commented on docs/store.js:
```diff
@@ -302,16 +299,24 @@ var firebaseCalendarStorage = function(params) {
+                var setActiveP = self.setActiveById(calendarObj.state.uniqueId);
+                var saveP = self.save(calendarObj);
+                Promise.all([addReaderP, setActiveP, saveP])
+                .then(function() {
```
I find that lining up the methods vertically if they are going to operate on the same object is an easy way to organize it.
```
Promise.all([...])
       .then(function() {
          ...
       })
       .catch(function() {});
```
However you decide to organize it, it should be such that lines that begin in the column indicate new *statements*, not expressions. If an expression continues onto a second line, it should be indented somehow to give a visual indication that this is all a single computation that will return a single value.

johntyree reviewed (changes requested):
This is probably good to go. I only had some cosmetic suggestions.

andrayantelo committed:
added back ticks and indentation
